python3Packages.psycopg: package pool and c variants

mweinelt opened the pull request:
Both implementations require the basic psycopg package and cannot exist
without it. Therefore they have no global name within the python package
set and are only accessible via psycopg's optional-dependencies.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
python3Packages.psycopg: package pool and c variants

Both implementations require the basic psycopg package and cannot exist
without it. Therefore they have no global name within the python package
set and are only accessible via psycopg's optional-dependencies.

risicle commented:
@ofborg eval

Unhappy with the metadata?

risicle commented on pkgs/development/python-modules/psycopg/c.nix:
```diff
@@ -0,0 +1,34 @@
+  inherit (psycopg) version;
+  format = "pyproject";
+
+  inherit (psycopg) src;
```
Worth also inheriting any `patches` psycopg has? Best to avoid using a mismatch of source and would be good to get any security patches we hurriedly add, forgetting about the existence of these leaf packages.

mweinelt replied:
Ah yes. Patching is a bit messy. Patches are best applied to the base repo, while `sourceRoot` needs to go one directory down for each project.

risicle replied:
Ugh forgot patching happens relative to `sourceRoot`.

risicle commented on pkgs/development/python-modules/psycopg/pool.nix:
```diff
@@ -0,0 +1,29 @@
+    typing-extensions
+  ];
+
+  doCheck = false;
```
Because this is tested by `psycopg`'s tests?

mweinelt replied:
Yes

risicle replied:
Might be worth including a comment to this effect.

risicle commented:
My macos builder certainly doesn't like the time-based `test_sched` tests either - possibly it's too slow?

risicle commented on pkgs/development/python-modules/psycopg/c.nix:
```diff
@@ -0,0 +1,34 @@
+    description = "PostgreSQL database adapter for Python -- C optimisation distribution";
+    homepage = "https://github.com/psycopg/psycopg/tree/master/psycopg_c";
+    license = licenses.lgpl3Plus;
+    maintaienrs = with maintainers; [ hexa ];
```
Typo spotted!

risicle commented on pkgs/development/python-modules/psycopg/pool.nix:
```diff
@@ -0,0 +1,29 @@
+    description = "Connection Pool for Psycopg";
+    homepage = "https://github.com/psycopg/psycopg/tree/master/psycopg_pool";
+    license = licenses.lgpl3Plus;
+    maintaienrs = with maintainers; [ hexa ];
```
Typo again.

mweinelt replied:
Oh no! Fixed :laughing:

mweinelt commented:
I dropped the extra files and move everything into the `default.nix`. Patching now affects all packages.

mweinelt commented:
Two more typos squashed.

```
-    # apply patches to base rpeo
+    # apply patches to base repo
     inherit patches;
```

risicle commented:
@ofborg build pgcli

risicle commented:
Yup darwin on ofborg has the same dislike as my darwin machine for the `test_sched` tests.

mweinelt commented:
Hrm, x86_64-darwin clearly fails a test, but aarch64-darwin fails the pythonImportsCheck.

mweinelt committed:
python3Packages.psycopg: disable test_sched{,_error} on darwin

They are prone to race conditions.

```
_______________________________ test_sched_error _______________________________
caplog = <_pytest.logging.LogCaptureFixture object at 0x197218a90>
    @pytest.mark.slow
    def test_sched_error(caplog):
        caplog.set_level(logging.WARNING, logger="psycopg")
        s = Scheduler()
        t = Thread(target=s.run, daemon=True)
        t.start()
        results = []
        def worker(i):
            results.append((i, time()))
        def error():
            1 / 0
        t0 = time()
        s.enter(0.1, partial(worker, 1))
        s.enter(0.4, None)
        s.enter(0.3, partial(worker, 2))
        s.enter(0.2, error)
        t.join()
        t1 = time()
        assert t1 - t0 == pytest.approx(0.4, 0.1)
        assert len(results) == 2
        assert results[0][0] == 1
>       assert results[0][1] - t0 == pytest.approx(0.1, 0.1)
E       assert 0.11003494262695312 == 0.1 ± 1.0e-02
E         comparison failed
E         Obtained: 0.11003494262695312
E         Expected: 0.1 ± 1.0e-02
tests/pool/test_sched.py:93: AssertionError
```

mweinelt commented:
@ofborg build pgcli

risicle reviewed (approved):
`nixpkgs-review` happy, macos 10.15, nixos x86_64 & aarch64
